keiko-redis/src/main/kotlin/com/netflix/spinnaker/q/redis/RedisQueue.kt

```diff
@@ -376,9 +377,34 @@ class RedisQueue(
   private fun JedisCommands.zismember(key: String, member: String) =
     zrank(key, member) != null
 
-  private fun Message.hash() =
+  private fun JedisCommands.anyZismember(key: String, members: Set<String>) =
+    members.any { zismember(key, it) }
+
+  private fun JedisCommands.firstFingerprint(key: String, fingerprint: Fingerprint) =
+    fingerprint.all.firstOrNull { zismember(key, it) }
+
+  @Deprecated("Hashes the attributes property, which is mutable")
+  private fun Message.hashV1() =
     Hashing
       .murmur3_128()
       .hashString(toString(), Charset.defaultCharset())
       .toString()
+
+  private fun Message.hashV2() =
```

Comment:
Can’t you just hash `message.payload` rather than converting back & forth?

Reply:
Haha derp. Yes

Reply:
It’s not that straightforward as the payload is not part of the `Message` interface. I wonder if we can expose that or include a hash function on the interface.

Reply:
I wouldn’t want every implementation to have to redefine the hash. It’s too error-prone. There’s not necessarily just a single `payload` property, though.

Reply:
One possibility is we add another layer to group properties on each implementation under a `payload` property. Not sure how painful a refactor that would be.

Reply:
This is why we should've had an actual wrapper class. :'(

Reply:
For on-lookers, we synced IRL: Going to keep with the object mapper route, as it's safer when mixing with Java POJOs or people overriding `toString`. We'll be looking to add message envelopes in a separate PR, sometime down the line, as this would just be a quality of life improvement for our own internal dev of keiko.